feat: OAuth2 controller api v2 + refactor oAuth Trpc handlers

ThyMinimalDev opened the pull request:
Fixes #26470 

## What does this PR do?

Adds a new OAuth2 controller for API v2 to support migrating OAuth endpoints from Next.js to the NestJS API. This includes **fully functional implementations** of client validation, PKCE verification, and JWT token generation/verification.

New endpoints created:
- `GET /v2/auth/oauth2/clients/:clientId` - Returns OAuth client info (mirrors `getClient.handler.ts`)
- `POST /v2/auth/oauth2/clients/:clientId/authorize` - Generates authorization code and **redirects** (mirrors `generateAuthCode.handler.ts`)
- `POST /v2/auth/oauth2/clients/:clientId/exchange` - Exchanges code for tokens (mirrors `token/route.ts`)
- `POST /v2/auth/oauth2/clients/:clientId/refresh` - Refreshes access token (mirrors `refreshToken/route.ts`)

Files created:
- Controller: `apps/api/v2/src/modules/auth/oauth2/controllers/oauth2.controller.ts`
- Service: `packages/features/oauth/services/OAuthService.ts` (shared for tRPC and API v2)
- Repositories: `packages/features/oauth/repositories/OAuthClientRepository.ts`, `AccessCodeRepository.ts`
- Input DTOs: `authorize.input.ts`, `exchange.input.ts`, `refresh.input.ts`
- Output DTOs: `oauth2-client.output.ts`, `oauth2-tokens.output.ts`
- Module: `oauth2.module.ts`
- E2E Tests: `oauth2.controller.e2e-spec.ts`
- DI modules: `packages/features/oauth/di/` (tokens, modules, container)

### Updates since last revision

**Latest changes (PR comment fixes):**
- Removed unused `GetOAuth2ClientInput` class
- Changed `@ApiTags("Auth / OAuth2")` to `@ApiTags("OAuth2")` for cleaner API organization
- Moved `OAuthClientFixture` to separate fixture file (`test/fixtures/repository/oauth2-client.repository.fixture.ts`)
- Added `type` property to `OAuth2ClientDto` to expose confidential/public client type
- Renamed `clientId` to `id` in `OAuth2ClientDto` output (using `@Expose({ name: "clientId" })` mapping)
- Cleaned up duplicate provider declarations in `oauth2.module.ts`
- Updated e2e tests to use new fixture and verify `type` property

**Previous changes (tRPC handler fix):**
- Fixed `generateAuthCode.handler.ts` to check for `ErrorWithCode` instead of `HttpError` - the handler was still checking for `HttpError` in its catch block, but `OAuthService` now throws `ErrorWithCode`, causing error messages to be lost and replaced with "server_error"

**Earlier updates:**
- Replaced `HttpError` with `ErrorWithCode` throughout OAuth files for consistent error handling
- Updated `OAuthService.ts` to throw `ErrorWithCode` with appropriate `ErrorCode` enum values
- **Authorize endpoint returns HTTP 303 redirect** with authorization code and state in query params
- Made `redirectUri` a **required** input parameter (exact match validation)
- Fixed P0 critical bug: `token_type` field name mismatch in `DecodedRefreshToken` interface
- Added `@UseGuards(ApiAuthGuard)` to `getClient` and `authorize` endpoints

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). N/A - new internal API endpoints hidden from public docs
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

1. **E2E Tests** - Run `TZ=UTC yarn jest --config ./jest-e2e.ts --testPathPattern="oauth2.controller.e2e-spec"` in `apps/api/v2`
2. **Unit Tests** - Run `TZ=UTC yarn vitest run packages/trpc/server/routers/viewer/oAuth/generateAuthCode.handler.test.ts`
3. **Redirect behavior** - Verify authorize endpoint returns 303 with Location header containing `code` and `state` params
4. **Error redirects** - Verify errors redirect to redirect URI with `error` and `error_description` query params
5. **DI wiring** - Verify `getOAuthService()` from `packages/features/oauth/di/OAuthService.container.ts` returns a properly instantiated service

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings

---

### Human Review Checklist
- [x] **IMPORTANT**: `getClient.handler.ts` still checks for `HttpError` but `OAuthService` now throws `ErrorWithCode` - verify this doesn't cause error messages to be lost (may need same fix as `generateAuthCode.handler.ts`)
- [x] Verify E2E tests pass with the `getToken` mock approach (mocks `next-auth/jwt` for `ApiAuthStrategy`)
- [x] Verify authorize endpoint returns HTTP 303 redirect (not JSON)
- [x] Verify `redirectUri` is required and exact match validation works correctly
- [x] Verify no debug `console.log` statements remain in production code
- [x] Review `as unknown as PrismaClient` type casting in `prisma-access-code.repository.ts` and `prisma-oauth-client.repository.ts`
- [x] Verify `ErrorWithCode` is properly exported from platform-libraries and error handling correctly maps `ErrorCode` to HTTP status codes via `getHttpStatusCode`
- [x] Verify `mapErrorToOAuthError` correctly handles `ErrorWithCode` instances (checks `error.code` instead of `error.statusCode`)
- [x] Confirm `@ApiExcludeController(true)` correctly hides endpoints from Swagger docs
- [ ] Verify `@Expose({ name: "clientId" })` correctly maps `id` property to `clientId` in JSON output
- [ ] Verify `type` property correctly exposes `clientType` enum value in OAuth2ClientDto

Link to Devin run: https://app.devin.ai/sessions/35eff6e6fcb340eea1ecaec74bf006f7
Requested by: Volnei Munhoz (volnei@cal.com) (@volnei)

devin-ai-integration[bot] committed:
feat(api-v2): add OAuth2 controller skeleton for auth module

- Add new OAuth2 module under /auth with controller, service, repository, and DTOs
- Implement skeleton endpoints:
  - GET /v2/auth/oauth2/clients/:clientId - Get OAuth client info
  - POST /v2/auth/oauth2/clients/:clientId/authorize - Generate authorization code
  - POST /v2/auth/oauth2/clients/:clientId/exchange - Exchange code for tokens
  - POST /v2/auth/oauth2/clients/:clientId/refresh - Refresh access token
- Create input DTOs for authorize, exchange, and refresh operations
- Create output DTOs for client info, authorization code, and tokens
- Register OAuth2Module in endpoints.module.ts

Co-Authored-By: morgan@cal.com <morgan@cal.com>

ThyMinimalDev committed:
chore: add missing functions to platform libraries

devin-ai-integration[bot] committed:
feat(api-v2): integrate OAuthService from platform-libraries into OAuth2 controller

- Add OAuthService export to platform-libraries
- Refactor OAuth2Service to use OAuthService.validateClient() and OAuthService.verifyPKCE()
- Remove duplicate local implementations of validateClient and verifyPKCE methods

Co-Authored-By: morgan@cal.com <morgan@cal.com>

devin-ai-integration[bot] committed:
feat(api-v2): use local validateClient and verifyPKCE implementations

- Remove OAuthService export from platform-libraries (will be deprecated)
- Reimplement validateClient and verifyPKCE methods locally in OAuth2Service
- Use verifyCodeChallenge and generateSecret from platform-libraries directly

Co-Authored-By: morgan@cal.com <morgan@cal.com>

ThyMinimalDev commented on apps/api/v2/src/modules/auth/oauth2/services/oauth2.service.ts:
```diff
@@ -0,0 +1,218 @@
+
+    let teamId: number | undefined;
+    if (teamSlug) {
+      const team = await this.oauth2Repository.findTeamBySlugAndUserId(teamSlug, userId);
```
DevinAI, this should be coming from the team repository

ThyMinimalDev commented on apps/api/v2/src/modules/auth/oauth2/services/oauth2.service.ts:
```diff
@@ -0,0 +1,218 @@
+      scopes,
+      codeChallenge,
+      codeChallengeMethod,
+    });
```
DevinAI, this should come from access code repository

ThyMinimalDev commented on apps/api/v2/src/modules/auth/oauth2/oauth2.repository.ts:
```diff
@@ -0,0 +1,179 @@
+    const randomBytesValue = randomBytes(40);
+    return randomBytesValue.toString("base64").replace(/=/g, "").replace(/\+/g, "-").replace(/\//g, "_");
+  }
+}
```
DevinAI, this should be in oauth2.service.ts, only prisma related code should be in repository

devin-ai-integration[bot] committed:
refactor(api-v2): separate repositories for OAuth2, access codes, and teams

- Create AccessCodeRepository for access code Prisma operations
- Add findTeamBySlugWithAdminRole method to TeamsRepository
- Move generateAuthorizationCode, createTokens, verifyRefreshToken to OAuth2Service
- Update OAuth2Repository to only contain OAuth client operations
- Update OAuth2Module to import TeamsModule and provide AccessCodeRepository

Addresses PR review comments to properly separate concerns

Co-Authored-By: morgan@cal.com <morgan@cal.com>

devin-ai-integration[bot] committed:
feat(api-v2): implement JWT token creation and verification for OAuth2

- Implement createTokens method using jsonwebtoken to sign access and refresh tokens
- Implement verifyRefreshToken method to verify JWT refresh tokens
- Add ConfigService injection for CALENDSO_ENCRYPTION_KEY access
- Import ConfigModule in OAuth2Module for dependency injection

Based on logic from apps/web/app/api/auth/oauth/token/route.ts and
apps/web/app/api/auth/oauth/refreshToken/route.ts

Co-Authored-By: morgan@cal.com <morgan@cal.com>

ThyMinimalDev committed:
Merge branch 'main' into devin/oauth2-controller-skeleton-1765988792

ThyMinimalDev committed:
fix: refactor and dayjs import fix

devin-ai-integration[bot] committed:
feat(api-v2): add ApiAuthGuard to getClient endpoint and e2e tests for OAuth2

- Add @UseGuards(ApiAuthGuard) to getClient endpoint for authentication
- Add comprehensive e2e tests for all OAuth2 endpoints:
  - GET /v2/auth/oauth2/clients/:clientId
  - POST /v2/auth/oauth2/clients/:clientId/authorize
  - POST /v2/auth/oauth2/clients/:clientId/exchange
  - POST /v2/auth/oauth2/clients/:clientId/refresh
- Tests cover happy paths and error cases (invalid client, invalid code, etc.)

Co-Authored-By: morgan@cal.com <morgan@cal.com>

devin-ai-integration[bot] committed:
chore(api-v2): hide OAuth2 endpoints from Swagger documentation

Co-Authored-By: morgan@cal.com <morgan@cal.com>

ThyMinimalDev committed:
hide endpoints from doc

ThyMinimalDev committed:
Merge branch 'main' into devin/oauth2-controller-skeleton-1765988792

devin-ai-integration[bot] committed:
fix(api-v2): address PR feedback for OAuth2 controller

- Fix P0 critical bug: token_type field name mismatch in DecodedRefreshToken interface
- Add @Equals('authorization_code') validation to exchange.input.ts grantType
- Add @Equals('refresh_token') validation to refresh.input.ts grantType
- Add @IsNotEmpty() validation to get-client.input.ts clientId
- Add @Expose() decorators to all output DTOs for proper serialization
- Add security test assertion to verify clientSecret is not returned in response

Co-Authored-By: morgan@cal.com <morgan@cal.com>

devin-ai-integration[bot] committed:
feat(api-v2): implement redirect behavior for OAuth2 authorize endpoint

- Update authorize endpoint to return HTTP 303 redirect with authorization code
- Add exact match validation for redirect URI (security requirement)
- Implement error handling with redirect to redirect URI and error query params
- Add state parameter support for CSRF protection
- Update e2e tests to verify redirect behavior (303 status, Location header, error redirects)

Co-Authored-By: morgan@cal.com <morgan@cal.com>

ThyMinimalDev commented on apps/api/v2/src/modules/auth/oauth2/controllers/oauth2.controller.ts:
```diff
@@ -0,0 +1,185 @@
+        state: body.state,
+      });
+
+      res.redirect(303, errorRedirectUrl);
```
DevinAi add return

devin-ai-integration[bot] committed:
refactor(api-v2): address PR feedback for OAuth2 authorize endpoint

- Make redirectUri a required input parameter (remove optional fallback)
- Move buildRedirectUrl and mapErrorToOAuthError to oauth2Service
- Add return statements to res.redirect() calls
- Simplify controller by delegating redirect URL building to service

Co-Authored-By: morgan@cal.com <morgan@cal.com>

ThyMinimalDev committed:
Merge branch 'main' into devin/oauth2-controller-skeleton-1765988792

ThyMinimalDev committed:
wip move code outside of api v2

devin-ai-integration[bot] committed:
feat(oauth): wire up dependency injection for OAuthService and repositories

Co-Authored-By: morgan@cal.com <morgan@cal.com>

ThyMinimalDev committed:
refactor: oAuthService used in routes

ThyMinimalDev committed:
Merge branch 'main' into devin/oauth2-controller-skeleton-1765988792

ThyMinimalDev committed:
fix imports

devin-ai-integration[bot] committed:
fix(api-v2): return 404 for invalid client ID instead of redirect in authorize endpoint

Co-Authored-By: morgan@cal.com <morgan@cal.com>

devin-ai-integration[bot] committed:
fix(api-v2): fix E2E test URL paths and remove bootstrap() in authenticated section

- Remove bootstrap() call in authenticated section to match working test pattern
- Change URL paths from /api/v2/... to /v2/... in authenticated section
- Keep bootstrap() and /api/v2/... paths in unauthenticated section

Co-Authored-By: morgan@cal.com <morgan@cal.com>

devin-ai-integration[bot] committed:
fix(api-v2): mock getToken from next-auth/jwt for E2E tests

- Add jest.mock for next-auth/jwt getToken function
- Mock returns null for unauthenticated tests
- Mock returns { email: userEmail } for authenticated tests
- Remove withNextAuth helper since ApiAuthGuard uses ApiAuthStrategy
  which calls getToken directly, not NextAuthStrategy

Co-Authored-By: morgan@cal.com <morgan@cal.com>

ThyMinimalDev committed:
fix tests and code review

ThyMinimalDev committed:
cleanup generate secrets

ThyMinimalDev committed:
cleanup controller

ThyMinimalDev commented on packages/features/oauth/services/OAuthService.ts:
```diff
@@ -1,61 +1,384 @@
+    }
+
+    try {
+      const decoded = jwt.verify(refreshToken, secretKey) as DecodedRefreshToken;
```
@CarinaWolli could you check this

devin-ai-integration[bot] committed:
chore: remove console.log from OAuthService.validateClient

Co-Authored-By: morgan@cal.com <morgan@cal.com>

ThyMinimalDev commented on apps/api/v2/src/modules/auth/oauth2/outputs/oauth2-authorize.output.ts:
```diff
@@ -0,0 +1,31 @@
+import { ApiProperty } from "@nestjs/swagger";
```
DevinAI delete this file it's dead code

ThyMinimalDev commented on apps/web/app/api/auth/oauth/token/route.ts:
```diff
@@ -27,83 +23,39 @@ async function handler(req: NextRequest) {
+      }
+    );
+  } catch (err) {
+    console.log(err, err instanceof HttpError);
```
DevinAI remove console.log

ThyMinimalDev committed:
Update apps/web/app/api/auth/oauth/refreshToken/route.ts

Co-authored-by: cubic-dev-ai[bot] <191113872+cubic-dev-ai[bot]@users.noreply.github.com>

devin-ai-integration[bot] committed:
fix(api-v2): add bootstrap() to authenticated E2E tests and update URL paths to /api/v2/

Co-Authored-By: morgan@cal.com <morgan@cal.com>

devin-ai-integration[bot] committed:
Merge branch 'devin/oauth2-controller-skeleton-1765988792' of https://git-manager.devin.ai/proxy/github.com/calcom/cal.com into devin/oauth2-controller-skeleton-1765988792

devin-ai-integration[bot] committed:
Merge remote-tracking branch 'origin/devin/oauth2-controller-skeleton-1765988792' and remove console.log from token route

Co-Authored-By: morgan@cal.com <morgan@cal.com>

ThyMinimalDev committed:
chore: remove dead code

ThyMinimalDev committed:
refactor

ThyMinimalDev committed:
refactor: generateAuthCode trpc handler and getClient trpc handler

ThyMinimalDev committed:
Merge branch 'main' into devin/oauth2-controller-skeleton-1765988792

ThyMinimalDev committed:
remove pkce check for refreshToken endpoint

ThyMinimalDev committed:
Update packages/trpc/server/routers/viewer/oAuth/getClient.handler.ts

Co-authored-by: cubic-dev-ai[bot] <191113872+cubic-dev-ai[bot]@users.noreply.github.com>

ThyMinimalDev committed:
refactor: error handling

ThyMinimalDev committed:
refactor: error handling

ThyMinimalDev commented:
@volnei @CarinaWolli ready to be tested with Companion app

ThyMinimalDev committed:
refactor: error handling

ThyMinimalDev committed:
remove console log

ThyMinimalDev commented on packages/trpc/server/routers/viewer/oAuth/generateAuthCode.handler.ts:
```diff
@@ -17,89 +16,33 @@ type AddClientOptions = {
-    // Optional PKCE validation for CONFIDENTIAL clients
-    if (codeChallenge && (!codeChallengeMethod || codeChallengeMethod !== "S256")) {
+  try {
+    const oAuthClient = await oAuthService.getClient(clientId);
```
TODO: caller of this endpoint should always provide the redirect uri

fetching it from db to avoid breaking change

ThyMinimalDev committed:
provide redirectUri in generateAuthCodeHandler

ThyMinimalDev committed:
provide redirectUri in authorize view

hbjORbj reviewed (changes requested):
Great stuff, just 1 comment

devin-ai-integration[bot] committed:
refactor: replace HttpError with ErrorWithCode in OAuth files

- Update OAuthService to use ErrorWithCode instead of HttpError
- Update mapErrorToOAuthError to check ErrorCode instead of status codes
- Update token/route.ts and refreshToken/route.ts to use ErrorWithCode
- Add ErrorWithCode export to platform-libraries
- Use getHttpStatusCode to map ErrorCode to HTTP status codes

Co-Authored-By: morgan@cal.com <morgan@cal.com>

devin-ai-integration[bot] committed:
fix: update generateAuthCode handler to use ErrorWithCode instead of HttpError

The handler was still checking for HttpError in its catch block, but OAuthService
now throws ErrorWithCode. This caused the error messages to be lost and replaced
with 'server_error' instead of the specific error messages.

Co-Authored-By: morgan@cal.com <morgan@cal.com>

devin-ai-integration[bot] committed:
fix: update OAuth2 controller to use ErrorWithCode instead of HttpError

The controller was still checking for HttpError in its catch blocks, but OAuthService
now throws ErrorWithCode. This caused all errors to return 500 Internal Server Error
instead of the correct HTTP status codes (400, 401, 404).

Also added getHttpStatusCode export to platform-libraries.

Co-Authored-By: morgan@cal.com <morgan@cal.com>

devin-ai-integration[bot] committed:
fix: add explicit unknown type annotation to catch blocks in OAuth2 controller

Co-Authored-By: morgan@cal.com <morgan@cal.com>

devin-ai-integration[bot] committed:
fix: change NotFoundException to HttpException in authorize endpoint

Co-Authored-By: morgan@cal.com <morgan@cal.com>

devin-ai-integration[bot] committed:
fix: handle err with ErrorWithCode in authorize endpoint catch block

Co-Authored-By: morgan@cal.com <morgan@cal.com>

ThyMinimalDev committed:
Merge branch 'main' into devin/oauth2-controller-skeleton-1765988792

ThyMinimalDev committed:
fix: move errorWithCode

ThyMinimalDev committed:
fix: error code rfc

ThyMinimalDev committed:
fix: no need to handle errors there is a middleware

ThyMinimalDev committed:
refactor errors

ThyMinimalDev committed:
refactor errors

ThyMinimalDev committed:
refactor redirecturi and state from api

ThyMinimalDev committed:
Merge branch 'main' into devin/oauth2-controller-skeleton-1765988792

PeerRich committed:
Merge branch 'main' into devin/oauth2-controller-skeleton-1765988792

hbjORbj reviewed (dismissed):
Great stuff

PeerRich committed:
Merge branch 'main' into devin/oauth2-controller-skeleton-1765988792

supalarry commented on apps/api/v2/src/modules/auth/oauth2/controllers/oauth2.controller.e2e-spec.ts:
```diff
@@ -0,0 +1,355 @@
+}));
+const mockGetToken = nextAuthJwt.getToken as jest.MockedFunction<typeof nextAuthJwt.getToken>;
+
+class OAuthClientFixture {
```
We store fixtures in separate files

supalarry commented on apps/api/v2/src/modules/auth/oauth2/outputs/oauth2-client.output.ts:
```diff
@@ -0,0 +1,64 @@
+  })
+  @IsString()
+  @Expose()
+  clientId!: string;
```
Since we are returning information about just the client, shall we just have 'id' instead of 'clientId'? aka we are not prefixing 'name' with 'clientName'

supalarry commented on apps/api/v2/src/modules/auth/oauth2/outputs/oauth2-client.output.ts:
```diff
@@ -0,0 +1,64 @@
+
+import { SUCCESS_STATUS, ERROR_STATUS } from "@calcom/platform-constants";
+
+export class OAuth2ClientDto {
```
Can we also return whether the oauth client is confidential or public under a new "type" property ?

supalarry reviewed (dismissed):
Review:

devin-ai-integration[bot] committed:
Merge branch 'main' into devin/oauth2-controller-skeleton-1765988792

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

devin-ai-integration[bot] committed:
refactor: address PR comments for OAuth2 controller

- Remove unused GetOAuth2ClientInput class
- Change API tag from 'Auth / OAuth2' to 'OAuth2'
- Move OAuthClientFixture to separate fixture file (oauth2-client.repository.fixture.ts)
- Add 'type' property to OAuth2ClientDto for confidential/public client type
- Rename 'clientId' to 'id' in OAuth2ClientDto (using @Expose mapping)
- Clean up duplicate provider declarations in oauth2.module.ts
- Update e2e tests to use new fixture and verify type property

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

volnei committed:
Merge branch 'main' into devin/oauth2-controller-skeleton-1765988792
